beater/handlers.go

```diff
@@ -102,15 +107,64 @@ func frontendSwitchHandler(feSwitch bool, h http.Handler) http.Handler {
 		if feSwitch {
 			h.ServeHTTP(w, r)
 		} else {
-			sendStatus(w, r, 403, errForbidden)
+			sendStatus(w, r, http.StatusForbidden, errForbidden)
+		}
+	})
+}
+
+func rateLimitHandler(rateLimit int, h http.Handler) http.Handler {
+
+	cache, _ := lru.New(10000)
+
+	var deny = func(ip string) bool {
+		if rateLimit < 0 {
+			return false
+		}
+		if !cache.Contains(ip) {
+			cache.Add(ip, rate.NewLimiter(rate.Limit(rateLimit), rateLimit))
+		}
+		var limiter, _ = cache.Get(ip)
+
+		return !limiter.(*rate.Limiter).Allow()
+	}
+
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if deny(extractIP(r)) {
+			sendStatus(w, r, http.StatusTooManyRequests, errTooManyRequests)
+			return
 		}
+		h.ServeHTTP(w, r)
 	})
 }
 
+func extractIP(r *http.Request) (ip string) {
+	var remoteAddr = func() string {
+		addr := r.RemoteAddr
+		return strings.Split(addr, ":")[0]
+	}
+
+	var forwarded = func() string {
+		forwardedFor := r.Header.Get("X-Forwarded-For")
+		client := strings.Split(forwardedFor, ",")[0]
+		return strings.TrimSpace(client)
+	}
+
+	var real = func() string {
+		return r.Header.Get("X-Real-IP")
+	}
+
+	for _, s := range []string{remoteAddr(), forwarded(), real()} {
+		if s != "" {
+			ip = s
```

Comment:
this way you are overwriting the `ip` in every iteration. How about 
```
if s != ""{ 
  return s
}
```

Reply:
I have a big dislike for named return values because it often makes simple things complicated. Here's an alternative:

```golang
for _, s := range []string{real(), forwarded(), remoteAddr()} {
    if s != "" {
        return s
    }
}
return ""
```